src/plugins/kibana/public/settings/sections/indices/add_data_steps/pipeline_setup/lib/__tests__/pipeline.js

```diff
@@ -175,26 +311,126 @@ describe('processor pipeline', function () {
     });
 
     it('should not allow the bottom item to be moved down', function () {
-      const pipeline = new Pipeline(null, Processor);
-      pipeline.add({ value: 'item1' });
-      pipeline.add({ value: 'item2' });
-      pipeline.add({ value: 'item3' });
+      const pipeline = new Pipeline();
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+      const processorIds = getProcessorIds(pipeline);
 
       const target = pipeline.processors[2];
       pipeline.moveDown(target);
 
-      expect(pipeline.processors[0].value).to.be('item1');
-      expect(pipeline.processors[1].value).to.be('item2');
-      expect(pipeline.processors[2].value).to.be('item3');
+      expect(pipeline.processors[0].processorId).to.be(processorIds[0]);
+      expect(pipeline.processors[1].processorId).to.be(processorIds[1]);
+      expect(pipeline.processors[2].processorId).to.be(processorIds[2]);
+    });
+
+  });
+
+  describe('updateParents', function () {
+
+    it('should set the first processors parent to pipeline.input', function () {
+      const pipeline = new Pipeline();
+      pipeline.input = { foo: 'bar' };
+
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+
+      pipeline.processors.forEach(p => sinon.stub(p, 'setParent'));
+
+      pipeline.updateParents();
+
+      expect(pipeline.processors[0].setParent.calledWith(pipeline.input)).to.be(true);
+    });
+
+    it('should set non-first processors parent to previous processor', function () {
+      const pipeline = new Pipeline();
+      pipeline.input = { foo: 'bar' };
+
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+
+      pipeline.processors.forEach(p => sinon.stub(p, 'setParent'));
+
+      pipeline.updateParents();
+
+      expect(pipeline.processors[1].setParent.calledWith(pipeline.processors[0])).to.be(true);
+      expect(pipeline.processors[2].setParent.calledWith(pipeline.processors[1])).to.be(true);
+      expect(pipeline.processors[3].setParent.calledWith(pipeline.processors[2])).to.be(true);
+    });
+
+    it('should set pipeline.dirty', function () {
+      const pipeline = new Pipeline();
+      pipeline.updateParents();
+
+      expect(pipeline.dirty).to.be(true);
     });
 
   });
 
+  describe('getProcessorById', function () {
+
+    it('should return a processor when suppied its id', function () {
+      const pipeline = new Pipeline();
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+      pipeline.add(TestProcessor);
+      const processorIds = getProcessorIds(pipeline);
+
+      const actual = pipeline.getProcessorById(processorIds[2]);
+      const expected = pipeline.processors[2];
+
+      expect(actual).to.be(expected);
+    });
+
+    it('should throw an error if given an unknown id', function () {
```

Comment:
I don't think throwing an error is the right thing to do here. Every usage of getProcessorById will need to be wrapped in a try/catch. I think it would be better to just return undefined.